### ISSUE TITLE
Damaging database structure by unrelated changes when altering TEXT column

### ISSUE THREAD

**Issue by kolardavid** [HeidiSQL/HeidiSQL#1023]
When altering column attributes of TEXT type column, column type is changed resulting in involuntary change of DB structure.

## Steps to reproduce this issue

1. Create column of type TEXT and leave Length field empty - you need to have collation with variable char length (like utf8 or utf8mb4)
2. Column is created, HeidiSQL shows TEXT and length of 65535
3. Do any changes on column, like (dis)allow NULL, or change default value and click Save
4. Column is suddenly changed to MEDIUMTEXT of length 16777215

## Current behavior

This bug seems to be present for few months, when HeidiSQL started to show Length information for TEXT columns. Because UTF-8 has variable length, it seems that altering column to TEXT(65535) make sit MEDIUMTEXT, because you may need more bytes to safely store 65535 characters in UTF-8 charset.

This bug HAPPENS with every text type - TINYTEXT, TEXT, MEDIUMTEXT, where every time is on altering switched to next bigger type.

This bug DOES NOT HAPPEN with text columns of collation with 1-byte length characters like "ascii_bin".

## Expected behavior

Altering column should never change type, unless you explicitly do so.

## Possible solution

Return to original behavior which were present few months ago, where Length was not specified for TEXT columns.

## Environment

* **HeidiSQL version:** 
  11.0.0.5974 (and tens of previous builds I was trying during last few months, since Length for TEXT columns were introduced)
* **Database system and version:** 
  MariaDB 10.4.12
* **Operating system:** 
  Windows 10 (64-bit)

**Comment by ansgarbecker** [comment]
Problem seems to be the length is included in the `ALTER TABLE` statement. MariaDB 10.3 even throws an error, instead of changing the data type to `MEDIUMTEXT`:

```
ALTER TABLE `agency_types2`
	CHANGE COLUMN `Column 3` `Column 3` TEXT(65535) NOT NULL COLLATE 'utf8mb4_general_ci' AFTER `description`;
/* SQL Error (1265): Data truncated for column 'Column 3' at row 1 */
```

Think I should remove the length portion from text data types. That shouldn't be displayed anyway.

**Comment by kolardavid** [comment]
Thank you very much for fast response. Not sure about MDB 10.3, but I am using 10.4 and there is no error. It just silently changes. It also happens with other servers I use, including MySQL and Percona. I found lot of my tables across servers with crippled columns because of this bug. Luckily, it seems to have no downsides except changing table model and makes table bigger on disk.

**Comment by ansgarbecker** [comment]
Forget the error message, that was due to me setting "NOT NULL". The silent change is also with MariaDB 10.3
I just also realized that `TEXT` allows a custom length, so I think I cannot just leave the length away from the `ALTER` query. Only if the user did not touch it probably, or if it's the default 65535.

**Comment by ansgarbecker** [comment]
Think I got it - please update to the next build, available in half an hour, and retest.

**Comment by kolardavid** [comment]
I will. The issue with length being 65535 or any other user set - is that while TEXT allows up to 65535 length of bytes, it might actually be 16383 up to 65535 characters, with utf8mb4 encoding. You never know. So it depends on collation, if TEXT(40000) is legal or will be changed to MEDIUMTEXT.

Also Length on TEXT behaves differently than on other types. For example INT(255) is legal and works fine, still resulting in tradition 4-byte int type, TEXT(999999) not.

There is also difference between VARCHAR and TEXT, I was just trying that to check if I am right. If you will set VARCHAR(255) on utf8mb4, you actually create internally memory space of equivalent VARCHAR(1023) on ascii_bin. Because on VARCHAR, you actually limit number of characters which can have 1-4 bytes. So if you want to create VARCHAR, which occupies in database space of 1BYTE (real length)+DATA, you have to create VARCHAR(63). VARCHAR(64) up will already create 2BYTE length descriptor. HOWEVER - when you create TINYTEXT column, which is TINYTEXT(255), you are also saying how many chars should fit - so 255 * 4 bytes on utf8mb4, which does not fit. The trick in TEXT is, that TINYTEXT, unlike VARCHAR, can actually store 255 characters, which are represented in utf-8 by single byte.

TL;DR, while you want 1-byte length VARCHAR, you have to limit on 63 chars no matter if you store 1 or 4 byte characters - so check is aware of codepage compatible string length. However in TINYTEXT, you have 255 bytes available and is up to you, if you will store 255 x "a" or 63 x "😁". So it is limited by real strlen() not aware of codepage.

I hope you know what I mean :)

**Comment by ansgarbecker** [comment]
Of course, characters may consume one to four bytes, depending on the encoding. See the [RFC](https://tools.ietf.org/html/rfc3629#section-3):

> In UTF-8, characters ... are encoded using sequences of 1 to 4 octets (= bytes).

The length in TEXT/MEDIUMTEXT/VARCHAR etc. is documented as the "_character length_", but the MySQL documentation is wrong - later it contradicts itself:

> ... the effective maximum length is less if the value contains multi-byte characters

So the length is counted in bytes, not in characters.

The silent data type change is therefore caused by the server, which multiplies the given length with the maximum number of octets/bytes per character, according to the column's charset. If you fire some `ALTER TABLE  ... TEXT(65535)`, then the server calculates 65535 * 4, and uses MEDIUMTEXT instead. That's why you don't see the issue in binary columns, which always have 1 byte per "character":

> An optional length M can be given for this type. If this is done, MySQL creates the column as the smallest TEXT type large enough to hold values M characters long.

Whatever, the issue in HeidiSQL is fixed, isn't it?
